core/lib/types/src/l1_batch_committer.rs

```diff
@@ -0,0 +1,134 @@
+use zksync_basic_types::{ethabi::Token, U256};
+
+use crate::{commitment::L1BatchWithMetadata, utils};
+
+pub trait L1BatchCommitter
+where
+    Self: std::fmt::Debug + Send + Sync,
+{
+    fn l1_commit_data(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> Token;
+    fn l1_commit_data_size(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> usize {
+        crate::ethabi::encode(&[Token::Array(vec![
+            self.l1_commit_data(l1_batch_with_metadata)
+        ])])
+        .len()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RollupModeL1BatchCommitter {}
+
+#[derive(Debug, Clone)]
+pub struct ValidiumModeL1BatchCommitter {}
+
+impl L1BatchCommitter for RollupModeL1BatchCommitter {
+    fn l1_commit_data(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> Token {
+        println!("RollupModeL1BatchCommitter");
+        let commit_data = if l1_batch_with_metadata
+            .header
+            .protocol_version
+            .unwrap()
+            .is_pre_boojum()
+        {
+            preboojum_l1_commit_data(l1_batch_with_metadata)
+        } else {
+            rollup_mode_l1_commit_data(l1_batch_with_metadata)
+        };
+        Token::Tuple(commit_data)
+    }
+}
+
+impl L1BatchCommitter for ValidiumModeL1BatchCommitter {
+    fn l1_commit_data(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> Token {
+        println!("ValidiumModeL1BatchCommitter");
+        let commit_data = if l1_batch_with_metadata
+            .header
+            .protocol_version
+            .unwrap()
+            .is_pre_boojum()
+        {
+            preboojum_l1_commit_data(l1_batch_with_metadata)
```

Comment:
hmm is it right that we silently use rollup data for pre-bojum? I understand it should never invoked, but if so let's be explicit and panic

Reply:
We discussed this in a meeting with Danil and we're not sure if it was relevant. Let us know what should be done there. 

Also, I'm not sure about panicking there, maybe we can handle the case in some other way.

Reply:
but, what other way? old protocol versions don't support validium in any way, so one must not attempt running it in this configueration. If this is done, we want to fail as early as possible - hence panic.

@Deniallugo is there something I'm missing?

Reply:
We'll try to detect this case in compile time, but if it's not possible we'll come back to the panic idea.